### PR TITLE
[Serverless] Add distributor action

### PR DESCRIPTION
**Pull request by AlCutter**
Update serverless witness actions to be a distributor

Part of #464

### CODE REVIEW

File: serverless/deploy/github/distributor/combine_witness_signatures/internal/distributor/state.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+				ret[i] = combined[c].raw
+				continue nextCheckpoint
+			}
+			c++
+		}
+		// No checkpoints with at least i witness signatures
+		ret[i] = nil
+	}
+	return ret, nil
+
```
Comment (mhutchinson) [inline serverless/deploy/github/distributor/combine_witness_signatures/internal/distributor/state.go:95]:
```suggestion
```

**Review by mhutchinson — APPROVED** [review]

File: serverless/deploy/github/distributor/README.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
+# Deploying serverless checkpoint distributor on GitHub
+
+> :warning: This is very experimental.
+
+Similar to serverless logs, we can deploy a file-based witnessed checkpoint
+distributor on GitHub too.
+
+For more details on witnessing, and the various roles involved (including
+`distributor`), please see the [witness README](/google/trillian-examples/witness)
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply serverless/deploy/github/distributor/README.md]: hehe, too clever :)
done.

File: serverless/deploy/github/distributor/README.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+## Try it out yourself
+
+To try it out:
+
+1. Create a fresh github repo to contain a distributor, and clone locally.
+2. Initialise the distributor state:
+    1. we'll use a directory called `distributor` in our repo to
+       store the state files
+    2. TODO: config file
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply serverless/deploy/github/distributor/README.md]: I kinda TODONE it, linked to the example config

File: serverless/deploy/github/distributor/README.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+       ```
+
+3. Place the above github action configs into the `.github/workflows` directory in
+   your distributor repo, and commit that too.
+4. Push these commits up to github.
+
+Now you can raise "incoming checkpoint" PRs which drop cosigned checkpoints into the
+`distributor/logs/<log_id>/incoming` directory, whereupon the `witness_validator` action
+should check the contents. Once the "incoming checkpoint" PRs are merged you
+should see the `combine_witness_signs` action running in response (check the
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply serverless/deploy/github/distributor/README.md]: done

File: serverless/deploy/github/distributor/combine_witness_signatures/README.md [hunk 5]
```diff
@@ -19,50 +18,63 @@ Entities which rely on the contents of the log can thereby trust that the view o
...
+                {
+                        "ID": "test",
+                        "PublicKey": "github.com/AlCutter/serverless-test/log+28035191+AVtQ/9lW+g90rQY3+pODJvMQ8X/tTvh/EuvCDLSmUk4S"
+                }
+        ]
+}
+```
 
-PRs containing cosigned checkpoint files under the log's `.../witness` directory are
+PRs containing cosigned checkpoint files under the distributor's `logs/<logID>/incoming` directory are
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply serverless/deploy/github/distributor/combine_witness_signatures/README.md]: Ah yes, I've used ellipsis elsewhere to suggest that this is internal structure to a defined directory, so I've added that here too - ta!

File: serverless/deploy/github/distributor/combine_witness_signatures/internal/distributor/state.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+import (
+	"fmt"
+	"sort"
+
+	"github.com/google/trillian-examples/formats/checkpoints"
+	"github.com/google/trillian-examples/formats/log"
+	"golang.org/x/mod/sumdb/note"
+)
+
+// UpdateOpts contains the settings for the distributor update.
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply serverless/deploy/github/distributor/combine_witness_signatures/internal/distributor/state.go:27]: yeah, good call - modified the `UpdateState` comment to reflect that.

File: serverless/deploy/github/distributor/combine_witness_signatures/internal/distributor/state.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+	Witnesses            []note.Verifier
+	MaxWitnessSignatures uint
+}
+
+// UpdateState incorporates any incoming checkpoints into the distributor state.
+func UpdateState(state, incoming [][]byte, opts UpdateOpts) ([][]byte, error) {
+	// Handle changes in number of required CPs
+	requiredSlots := int(opts.MaxWitnessSignatures + 1)
+	if l := len(state); l > requiredSlots {
+		state = state[0:requiredSlots]
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply serverless/deploy/github/distributor/combine_witness_signatures/internal/distributor/state.go]: done

File: serverless/deploy/github/distributor/combine_witness_signatures/internal/distributor/state.go (hunk 1 above)
Comment (AlCutter) [reply serverless/deploy/github/distributor/combine_witness_signatures/internal/distributor/state.go]: done

**Pull request merged**